jdeeco-core/src/cz/cuni/mff/d3s/deeco/runtimelog/RuntimeLogger.java

```diff
@@ -0,0 +1,537 @@
+	 * for the logging.
+	 * @throws IllegalStateException Thrown if the {@link RuntimeLogger#init} method hasn't
+	 * been called before this method is being called.
+	 * @throws IllegalArgumentException Thrown if the given argument is null.
+	 */
+	public static void logSnapshot(RuntimeLogRecord snapshotData)
+			throws IOException, IllegalStateException {
+		log(snapshotData, true);
+	}
+
+	/**
+	 * Writes the data from the given <em>record</em> to the {@link RuntimeLogger#dataWriter}.
+	 * The written record has a format of an <a href="http://en.wikipedia.org/wiki/XML">XML</a>
+	 * element. The values from the given <em>record</em> argument are stored as attributes of the
+	 * written <a href="http://en.wikipedia.org/wiki/XML">XML</a> element. If the <em>snapshot</em>
+	 * argument is true also an index record pointing to the current snapshot record is written
+	 * using the {@link RuntimeLogger#indexWriter}.
+	 * <p> The {@link Writer}s are being flushed when the <em>snapshot</em> argument is true. </p> 
+	 * @param record contains the data that will be logged.
+	 * @param snapshot indicates whether the <em>record</em> holds a snapshot (true) or an event (false).
+	 * @throws IOException Thrown if there is a problem writing into the {@link Writer}s
+	 * for the logging.
+	 * @throws IllegalStateException Thrown if the {@link RuntimeLogger#init} method hasn't
+	 * been called before this method is being called.
+	 * @throws IllegalArgumentException Thrown if the given argument is null.
+	 */
+	private static void log(RuntimeLogRecord record, boolean snapshot) // TODO: check whether the record fields are not null
+			throws IOException, IllegalStateException {
+		if (instance.timeProvider == null)
+			throw new IllegalStateException(
+					"The SimLogger class not initialized.");
+		if (record == null)
+			throw new IllegalArgumentException(String.format(
+					"The argument \"%s\" is null.", "record"));
+
+		StringBuilder recordBuilder = new StringBuilder();
+		String recordName = snapshot ? SNAPSHOT_RECORD_NAME : EVENT_RECORD_NAME;
+		long currentDataOffset = instance.currentDataOffset;
+
+		recordBuilder.append("<").append(recordName).append(" ")
+				.append(RECORD_ID).append("=\"")
+				.append(record.getComponentID()).append("\" ")
+				.append(RECORD_TYPE).append("=\"")
+				.append(record.getEventType()).append("\" ")
+				.append(RECORD_TIME).append("=\"")
+				.append(instance.timeProvider.getCurrentMilliseconds())
+				.append("\" ");
+
+		Map<String, Object> values = record.getValues();
+		for (String valueKey : values.keySet()) {
+			recordBuilder.append(valueKey).append("=\"")
+					.append(values.get(valueKey)).append("\" ");
+		}
+
+		recordBuilder.append("/>\n");
+
+		try {
+			String compiledRecord = recordBuilder.toString();
+			instance.dataWriter.write(compiledRecord);
+			instance.currentDataOffset += compiledRecord.getBytes(CHARSET_NAME).length;
+
+			if (snapshot) {
```

Comment:
This is not gonna work because snapshots are not the only points to which one needs to go back. There are also the event which are registered in the backlog times.